app/bin/tools/datastore.dart

```diff
@@ -0,0 +1,335 @@
+                  ..created = versionArchive.created
+                  ..pubspec = new Pubspec(versionArchive.pubspecJson)
+                  ..readmeFilename = versionArchive.readmeFilename
+                  ..readmeContent = versionArchive.readmeContent
+                  ..changelogFilename = versionArchive.changelogFilename
+                  ..changelogContent = versionArchive.changelogContent
+                  ..exampleFilename = versionArchive.exampleFilename
+                  ..exampleContent = versionArchive.exampleContent
+                  ..libraries = versionArchive.libraries
+                  ..downloads = versionArchive.downloads
+                  ..sortOrder = versionArchive.sortOrder
+                  ..uploaderEmail = versionArchive.uploaderEmail;
+                pkgVersionUpdateCount++;
+                tx.queueMutations(inserts: [pv]);
+                await tx.commit();
+              } else {
+                await tx.rollback();
+              }
+            });
+          }
+
+          pool.withResource(updateVersion);
+        }
+        await Future.wait(futures);
+        await pool.close();
+      }
+      print('$pkgCounter packages processed from ${packagesFile.path}');
+      print('Updated:\n'
+          '  $pkgUpdateCount packages\n'
+          '  $pkgVersionUpdateCount versions.');
+    });
+  }
+
+  Future<String> _detectLatest(Directory backupDir) async {
+    final suffix = '-packages.jsonl';
+    final list = await backupDir
+        .list()
+        .where((fse) => fse is File && fse.path.endsWith(suffix))
+        .map((f) => f.path.substring(0, f.path.length - suffix.length))
+        .toList();
+    list.sort();
+    return list.last;
+  }
+}
+
+@JsonSerializable()
+class ArchiveLine {
+  final PackageArchive package;
+  final List<PackageVersionArchive> versions;
+
+  ArchiveLine({
+    @required this.package,
+    @required this.versions,
+  });
+
+  factory ArchiveLine.fromJson(Map<String, dynamic> json) =>
+      _$ArchiveLineFromJson(json);
+
+  Map<String, dynamic> toJson() => _$ArchiveLineToJson(this);
+}
+
+@JsonSerializable()
+class PackageArchive {
+  final String name;
+  final DateTime created;
+  final DateTime updated;
+  final int downloads;
+  final String latestVersion;
+  final String latestDevVersion;
+  final List<String> uploaderEmails;
+  final bool isDiscontinued;
+  final bool doNotAdvertise;
+
+  PackageArchive({
+    @required this.name,
+    @required this.created,
+    @required this.updated,
+    @required this.downloads,
+    @required this.latestVersion,
+    @required this.latestDevVersion,
+    @required this.uploaderEmails,
+    @required this.isDiscontinued,
+    @required this.doNotAdvertise,
+  });
+
+  factory PackageArchive.fromJson(Map<String, dynamic> json) =>
+      _$PackageArchiveFromJson(json);
+
+  Map<String, dynamic> toJson() => _$PackageArchiveToJson(this);
+}
+
+@JsonSerializable()
+class PackageVersionArchive {
```

Comment:
These classes need to stay in sync with the main classes. Can we avoid that? If not, add some comments to these classes and their counterparts telling the developer that they must stay in sync.

Reply:
I've added the notes. I could imagine a test that checks if they are in-sync, in one way or the other:
- making a common interface that both of these classes implement
- parsing the source files with analyzer and doing a comparison
- parsing the classes with dart:mirrors, although that may not be a good long-term solution (?)